experiments/lock.py

```diff
@@ -0,0 +1,146 @@
+
+import django
+from django.db import models, transaction
+from django.db.utils import IntegrityError, OperationalError
+from django.utils.encoding import python_2_unicode_compatible
+
+from experiments.dateutils import now
+
+
+logger = logging.getLogger(__file__)
+TIMEOUT_MAX = 9223372036.0
+
+
+@python_2_unicode_compatible
+class DbLock(models.Model):
+    """Simple expirable lock, backed by database"""
+    DEFAULT_TIMEOUT = 5  # seconds
+    RETRY_INTERVAL = 1  # seconds
+    DB_ERROR_RETRY_INTERVAL = 0.1  # seconds
+    name = models.SlugField(primary_key=True)
+    uuid = models.CharField(
+        max_length=36, null=False, unique=True, db_index=True,
+        default=uuid.uuid4)
+    expire_at = models.DateTimeField(null=False, db_index=True)
+
+    def __str__(self):
+        return '<{state} {name} object at {uuid}>'.format(
+            state='locked' if self.locked() else 'unlocked',
+            name=self.name,
+            uuid=self.uuid,
+        )
+
+    @transaction.atomic
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
+        if not self.uuid:
+            self.uuid = uuid.uuid4()
+        super(DbLock, self).save(
+            force_insert, force_update, using, update_fields)
+
+    @classmethod
+    def cleanup(cls):
+        """Delete expired locks."""
+        try:
+            cls.objects.filter(expire_at__lte=now()).delete()
+        except OperationalError:
+            pass
+
+    def acquire(self, blocking=True, timeout=None):
+        """
+        Acquire the lock.
+        If acquired, lock will be auto-release after <timeout> seconds.
+        If not acquired and <blocking>, will keep retrying every second
+        for <timeout> seconds.
+        """
+        if not blocking and timeout is not None:
+            raise ValueError('Cannot set timeout if not blocking')
+        timeout = self.DEFAULT_TIMEOUT if timeout is None else timeout
+        if timeout < 0:
+            raise ValueError('Cannot set negative timeout')
+        if timeout > TIMEOUT_MAX:
+            raise OverflowError('Timeout too large')
+        self.cleanup()
+        expire_at = now() + timedelta(seconds=timeout)
+        acquired = self._acquire(blocking, expire_at)
+        if acquired:
+            self.expire_at = now() + timedelta(seconds=timeout)
+            self.save()
+        return acquired
+
+    def _acquire(self, blocking, expire_at):
+        try:
+            lock, acquired = DbLock.objects.get_or_create(
+                name=self.name,
+                defaults={'expire_at': expire_at},
+            )
+        except (IntegrityError, OperationalError):
+            acquired = False
+            lock = None
+        if acquired:
+            self.uuid = lock.uuid
+            return True
+        if not blocking:
+            return False
+        while not acquired and now() < expire_at:
+            sleep(self.RETRY_INTERVAL * random.uniform(0.75, 1.25))
```

Comment:
Do the numbers in the `random.uniform` interval have a special meaning or were they chosen as a safe time frame for a retry?

Reply:
Just arbitrary random retry interval. Honestly, I was trying to get the standard Python tests for `Lock` class pass with SQLite but no luck, gave up in the end. I left these checks in anyway because I guess that other DBs could theoretically become locked. Although I don't think that happens realistically.

Reply:
Actually, the tests pass with SQLite. Just that in verbose mode we get console output for every `OperationalError` that is thrown. Never mind that it is handled with a retry, it still gets logged, and that looks like stuff is breaking. That's why I'm skipping some tests if `sqlite` DB backend is configured.